Add "watchable-mounts" concept to allow for  inotify support of specific types of mounts.

egernst opened the pull request:
Add "watchable-mounts" concept to allow for  inotify support of specific types of mounts.
    
  To workaround virtiofs' lack of inotify support, we'll special case
  particular mounts which are typically watched, and pass on information
  to the agent so it can ensure that the mount presented to the container
  is indeed watchable (see applicable agent commit).
    
  This PR will add docs and make the following functional changes:

-  Update runtime to:
    - identify watchable mounts based on file count and mount source
    - create a watchable-bind storage object for these mounts to
       communicate intent to the agent
    - update the OCI spec to take the updated watchable mount source into account
- Update agent to:
    - support watchable-bind storage driver
    - poll watchable-bind source in order to keep watchable-bind's mount point up to date

egernst commented:
/test

egernst commented:
/test

egernst commented:
/test

lifupan commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+        let dest_file_path = self.make_dest_path(&source_file_path)?;
+
+        if let Some(path) = dest_file_path.parent() {
+            fs::create_dir_all(path).await?;
```
The dest_file_path parent dir would exist, thus it should check the case and skip to create it again, otherwise, it would cause an error here.

lifupan commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+
+/// Represents a single storage entry (may have multiple files to watch).
+#[derive(Default, Debug, Clone)]
+struct Entry {
```
It's better to iml the Drop trait for Entry to do the clean up of the files in mount_point dir, otherwise, the trash files would be left after a container stopped.

lifupan commented on src/agent/src/rpc.rs:
```diff
@@ -253,11 +253,14 @@ impl AgentService {
-                .ok_or_else(|| anyhow!("Invalid container id"))?;
 
-            ctr.destroy().await?;
+            sandbox.bind_watcher.remove_container(&cid).await;
```
It's better to do the bind_watcher.remove_container after destroy the container, in case there's some cleanup in the bind watcher's volume depended on the container's destroy such as rootfs umount.

sameo commented on src/agent/src/netlink.rs:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2021 Kata Maintainers
+// Copyright (c) 2021 Apple Inc.
```
I don't think this one is related to that PR.

egernst replied:
agreed, we should drop.

jodh-intel commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+        let entry = File {
+            file: file.to_path_buf(),
+            // Set minimum value to perform initial copy during next check
+            modified: SystemTime::UNIX_EPOCH,
```
Won't this cause all files to be copied the first time `check()` is called? Why not just use the real file modification time here?

egernst replied:
We want the files to be copied over when we are setting up this "watch"

jodh-intel commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+use crate::protocols::agent::Storage;
+
+/// The maximum number of file system entries agent will watch for each mount.
+const MAX_ENTRIES_PER_STORAGE: usize = 8;
```
As mentioned, maybe this should be configurable with `8` being the default?

egernst replied:
I'm inclined to start opinionated with eight and make it configurable if/when we have a use case for it. This way we can keep it simple to start; I'm not sure we'll ever need to adjust this, and it'll avoid having to pass this information from runtime to agent, as well as the config scaffolding. WDYT? Am I just being lazy?

jodh-intel commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+const MAX_ENTRIES_PER_STORAGE: usize = 8;
+
+/// How often to check for modified files.
+const WATCH_INTERVAL_SECS: u64 = 2;
```
Another potential configurable (see `src/config.rs`). If configurable, milliseconds may be better.

egernst replied:
Same as above.

jodh-intel reviewed (commented):
Thanks @egernst and @mxpv. A few comments...

Also, although Kata isn't doing the inotify work, note that inotify has some rather dark corners (including the BUGS section in the man page and that "recursive watches" are racy and may result in DoS due to resource starvation).

Finally, can the agent still be shutdown cleanly with the code on this PR (using [`agent-ctl`](https://github.com/kata-containers/kata-containers/tree/main/tools/agent-ctl) specifying `-c DestroySandbox`).

egernst commented on docs/design/inotify.md:
```diff
@@ -0,0 +1,25 @@
+### Presenting a `watchable` mount to the workload
+
+For mounts that are considered `watchable`, inside the guest, the `kata-agent` will poll the mount presented from
+the host through `virtiofs` and copy any changed files to a `tmpfs` mount that is presented to the container. In this way,
```
Yeah, agreed.

In this case a bad actor would be in the namespace of the pod that is consuming the configmap, for example, and would be impacting just that single pod. Realistically I considered this to be someone shooting themselves in the foot, and worse case OOMing their containers.

egernst replied:
What do you think about adding some safe guards in the guest to limit tmpfs size to a sane limit (again, configmaps and certs/secrets are usually a single text file)?

If a file changes very often, that's okay. We only check for updates ever 2 sec (default today). If someone does this, we'll pick up the last change next cycle.

egernst replied:
I was thinking through this a bit more, and considering how to handle this safely, not failing silently, and not introducing too much extra complexity.

The initial implementation would check for the number of unique files in the configmap/secret at container creation time, and not enable polling (watcher) if the number is greater than 8. This was naive, though -- the number of files could increase/decrease over time (in particular for configmap, we'll actually add a file, then remove a file when the configmap is updated, so I still don't expect a large increase, but... expect the unexpected...). Similarly, right now we don't check the file(s) size, which as you point out could result in issues for the pod, since it'd be backed by tmpfs in the guest.

A suggestion: If we notice that the number of files or size of the files increase over a threshold for mount foo, we should:
- log an error
- remove the watcher / stop polling
- remove the contents at `MountPoint` of the watchable-bind storage object, and bind mount the storage objects `Source` over `MountPoint`

In this way, we stop polling, inotify stops working, but the contents of the mount are still available (and updated) in the container. This will match the functionality that exists today without the inotify handling.

WDYT?

egernst replied:
Updated docs to cover. this.

egernst commented on src/runtime/virtcontainers/mount.go:
```diff
@@ -446,13 +449,78 @@ func Isk8sHostEmptyDir(path string) bool {
+
+	for _, file := range files {
+		if file.IsDir() {
+			inc, err := countFiles(filepath.Join(path, file.Name()), (limit - numFiles))
```
Ah, let me dig more into this.

Usually these mounts would be very static, especially in file count, but, ... "usually...."

egernst replied:
This is an initial check before asking that the agent try to watch the mount, and I'd "expect" that normal path, this will be static, especially at container creation time.

Really, the mount can change at any time during container's life, so the agent  will check the total file count (and size) each time it polls, and stop polling in case.

egernst commented:
@jodh-intel thanks for the review. For the agent shutdown -- do we have an e2e test in our CI for this? If not, can we add? If it isn't in CI, I probably already broke it :D

egernst commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,487 @@
+}
+
+#[derive(Default, Debug)]
+struct Entries(Vec<Entry>);
```
An entry is created based on a storage objects source and destination, provided from the kata runtime. Based on this, these will always be unique within the sandbox, since the mount is renamed as $containerId-$uniqueID-$containerMountpoint.

egernst commented:
@jodh-intel et al -- I updated the PR to improve how we handle "too many files" and "too large of a mount" scenarios.

We now cap a mount being 'watchable' at 8 files, and /or 1MB of total size. The total size aligns with limit on [ConfigMap upstream ](https://kubernetes.io/docs/concepts/configuration/configmap/#motivation), and 8 is a sane number. When we see this occur, we stop treating this as a watchable bind, and instead just treat as standard bind mount. I updated the docs to also reflect this.

PTAL! I think this is in good shape for further review now.

egernst commented:
/test

fidencio commented:
/retest-ubuntu
/retest-metrics

fidencio commented:
@egernst, I went through the patches and they do look good to me.
So, for now, take this as a LGTM but I'm not approving them as I don't want to bulldoze / bypass @jodh-intel's or @lifupan's reviews.

Thanks!

egernst commented:
/retest-ubuntu
/retest-metrics

mxpv commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+
+/// Represents a single watched storage entry which may have multiple files to watch.
+#[derive(Default, Debug, Clone)]
+struct WatchedStorage {
```
nit: I'd suggest to avoid module prefixes in module impl, so you won't end up with `watcher::WatchableStorage`, but `watcher::Storage`. It's not a deal breaker here, just more rusty: https://rust-lang.github.io/rfcs/0356-no-module-prefixes.html

mxpv commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+    // Flag to indicate that the Storage should be watched.
+    watch: bool,
+
+    // The list of files to watch from the source mount point and updated in the target one.
```
nit: If we ever decide to generate rustdocs for kata crates, it's better to use `///` instead of `//`. This way it'll end up in docs.

mxpv reviewed (approved):
Just a couple of nits, which are not blockers. LGTM

sameo commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+#[derive(Debug, Default)]
+pub struct BindWatcher {
+    /// Container ID -> Vec of watched entries
+    shared: Arc<Mutex<HashMap<String, Entries>>>,
```
I'm really picking on naming, sorry about that. Using `shared` makes the code below harder to read, while renaming it to e.g `containers_storage` would make e.g. the following code:

```Rust
self.containers_storage
            .lock()
            .await
            .entry(id)
            .or_insert_with(ContainerStorage::default)
            .add(mounts, logger)
            .await
            .with_context(|| "Failed to add container")?;
```

easier to understand/read.

sameo commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+            self.mount(logger).await?;
+
+            // Spawn background thread to monitor changes
+            let join_handle = Self::spawn_watcher(
```
nit: 
```Rust
self.watch_thread = Some(Self::spawn_watcher(
...
```

egernst replied:
The cost of variables and lines in rust land must be hella expensive :D

sameo replied:
The Rust toolchain makes you so dumb that you want to pretend you still know better than LLVM and want to help it. I'm sure I'm saving one register renaming here, it's crucial.

sameo reviewed (commented):
A lot of nitpicking and renaming requests. The code looks pretty good overall, this is some relatively heavy lift to work around a virtiofs/FUSE limitation.

egernst commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+}
+
+#[derive(Default, Debug)]
+struct Entries(Vec<WatchedStorage>);
```
It needn't necessarily be tied to a single container. Perhaps SandboxStorage?

sameo replied:
`SandboxStorages` sounds good to me.

egernst commented on src/agent/src/watcher.rs:
```diff
@@ -0,0 +1,779 @@
+                continue;
+            }
+
+            match entry.scan(logger).await {
```
Oh interesting.  I started to take a look.  Any chance you can add a PR (that I'll squash) showing how you'd change it?

egernst replied:
I'm a bit anxious to get this landed -- are you okay if we do this nit in a follow on PR?

sameo replied:
Fine with me. Could we at least remove the 

```Rust
if !entry.watch
```
block with a `filter` on the `iter_mut()` ?

sameo replied:
Oh, and as @mxpv suggested, could we simply do:

```Rust
if let Err(e) = entry.scan(logger).await {
   ...error block...
}
```

and avoid having to artificially handle the `Ok` case?

sameo replied:
See https://github.com/egernst/kata-containers/pull/3 for that.

egernst replied:
Thanks -- done!

egernst commented:
/test

egernst commented:
/test

egernst commented:
/retest-ubuntu
/retest-fedora

egernst commented:
/test

egernst commented:
/retest-ubuntu
/retest-debian

fidencio commented:
/retest-arm
/retest-ubuntu

lifupan commented on src/agent/src/rpc.rs:
```diff
@@ -253,11 +253,14 @@ impl AgentService {
+
+            sandbox
+                .get_container(&cid)
+                .ok_or_else(|| anyhow!("Invalid container id"))?
```
Hi @egernst 
It seemed you missed those two comments?
